### ISSUE TITLE
Add a command to force color preview in the current view

### ISSUE THREAD

**Issue by gwenzek** [facelessuser/ColorHelper#103]
Thanks for this great plugin !

I have a Json file that contains some Hex colors.

I don't want to add a "color_scanning" rule for all JSON files, because most of the time my JSON files doesn't contain colors.

I'd like to be able either to:

* whitelist this specific file (eg by using a field "whitelisted_files" in a "color_scanning" conf)
* run a command from the palette that looks for colors in the current file and show the popups.

**Comment by FichteFoll** [comment]
Same here with a YAML file. Colors are specified like `'0x9ec400'` and I'd like to see previews and edit those, but I can't even insert a color when I open the picker via the command palette, "enter new color" and modify that. The popup just closes.

**Comment by facelessuser** [comment]
@FichteFoll, the insertion is a separate issue that should be tracked separately (with that said, I haven't bothered to reproduce the issue yet).

The idea with ColorHelper is to target specific scopes with a high chance of having a valid color.  If we just force scan a file without rules specified, we would just scan source and text which would include everything.  Keys, comments, and other things that wouldn't not normally have color, would get scanned.  Possibly variables names that happen to match web color names might show previews.  Also, what color formats are valid? Your JSON file might not allow all CSS3 colors, maybe just hex.

So, rules are required, but we could maybe allow a way to have disabled rules who can be turned on when you need them.

**Comment by gwenzek** [comment]
Could we have a command that take the rules as arguments? So I could add a
menu entry for searching hex colors in string. (Maybe that menu entry could
be added by default, that seems like a reasonable use case)

**Comment by facelessuser** [comment]
It's a possibility.  I support, honestly, too many projects, so ones I don't use as much get the least amount of attention.

I'm a firmware programmer, not a web developer, so ColorHelper isn't something that gets my full attention.  It was mainly a "cool" idea I wanted to run with for fun, and a lot of people have found it helpful.  But when it comes down to the core plugins I use, ColorHelper trails near the end as it isn't relevant to my daily work.

So, while I'll say, yeah, maybe that could work, I don't know when I'll actually get around to playing with it and making the final decision. So yeah, maybe 🙂.

**Comment by FichteFoll** [comment]
I'm not a webdev either, but I do sometimes work with color specifications (like for my terminal emulator, this time, or for ST color schemes).

I realize that this isn't top priority for you, but wanted to provide some input to the issue regardless.

My imagination regarding usage pattern of a feature like this would be to add a command to explicitly scan the current view (or selection?) for some recognized color patterns. These color patterns should be configurable in some way, either as arguments to the command or in the main config (or both). The advantage of the latter is that I can just run the default command from the command palette and have it fall back to the default configuration I set for my use case, which would not involve CSS3 colors or rgb and hsl function calls, btw.

I figured out reproduction steps for the separate issue and will create a new one for that.

**Comment by gwenzek** [comment]
Thanks to an insomnia, I made a PR adding this feature: #105

**Comment by AndydeCleyre** [comment]
I'm adding a voice to say I would love to just force color checking on arbitrary opened files. It would help me in debugging (figuring out if I have a bad configuration or some plugin bug), as the hover or inline or other color info doesn't automatically show for me for any file at all, though I can right-click and choose `Color Helper` > `Color Info` to get a popup. Additionally it would be nice for the odd json/yaml/template file.

Maybe off-topic now, but FWIW this is my conf, just trying to get this to work on any file:

```json
{
    "color_scanning": [
        {
            "syntax_filter": "blacklist",
            "base_scopes": [],
        }
    ]
}
```

**Comment by facelessuser** [comment]
You can do something like:

```js
{
    "color_scanning": [
        {
            "syntax_filter": "blacklist",
            "base_scopes": ["source", "text"],
            "scan_scopes": ["source", "text"],
            "allowed_colors": ["css3"]
        }
    ]
}
```

Beware in large files. I need to one day rewrite how we search for colors to make it perform better.
This whole thing started as a fun little experiment and proof of concept, but the scanning needs to be improved.

**Comment by facelessuser** [comment]
So don't complain about the performance of this if it goes south :wink:.

**Comment by AndydeCleyre** [comment]
Thank you, that's great! It's much better for me to have it always parse the open file, and use package control to enable or disable the plugin as needed. This is super handy, thank you.

**Comment by facelessuser** [comment]
The current plan is to provide a generic option. By default, scanning will be disabled. It will use the new default color class and it will allow users to use the color picker and the other new color tools and insert those colors back into the documentation. You can manually open the color info dialog if desired, or color picker, or anything else.

If you want the previews, you'd have to enable `allow_scanning`. It is likely we can create a toggle for a specific file view so it can have scanning manually turned on/off for just that view.

If it is found that certain colors (like maybe hex codes) are problematic for the user, they can remove the hash pattern from `color_trigger`. If they only want to scan certain color spaces, they can filter out entire color spaces with the `filter` option.

Triggers are just so we can do a quick search to find where we should check for a potential color. If the user needs to alter the recognized input for one or more color spaces and/or modify the currently known output formats, they'll have to subclass the default color class and provide the import path via `color_class`. That is, of course, more advanced.

Generic will only kick in when no other file rule is found. And scanning will only occur when `allow_scanning` is enabled as well.

Something similar will be possible via per file rules as well. The main difference between those and Generic, is that generic has no file extension, syntax file, or other options. It basically provides compatibility so users can use ColorHelper anywhere they see a recognized color.

```js
    "generic": {
        "allow_scanning": false,
        "color_trigger": "(?i)(?:\\b(?:color|hsla?|gray|lch|lab|hwb|rgba?)\\(|\\b(?<!\\#)[\\w]{3,}(?!\\()\\b|\\#)",
        "color_class": "coloraide.css.Color",
        "scan_scopes": ["-comment -string"],
        "output_options": [
            {"space": "srgb", "format": {"options": {"hex": true}}},
            {"space": "srgb", "format": {"options": {"comma": true}, "precision": 0}},
            {"space": "hsl", "format": {"options": {"comma": true}, "precision": 0}},
            {"space": "hwb", "format": {"options": {"comma": false}, "precision": 0}},
            {"space": "lch", "format": {"options": {"comma": false}, "precision": 0}},
            {"space": "lab", "format": {"options": {"comma": false}, "precision": 0}},
            {"space": "xyz", "format": {}}
        ]
    },
```

Thinks are still a bit in flux, but this should have a path forward in the next major release coming.

**Comment by facelessuser** [comment]
Assuming the user has a generic fallback option set with some scan scopes, the user will be able to manually, per view, enable color scanning via a new override command. This is the default generic, so it will scan anything except comments and strings. It assumes the default output options and triggers etc.

This will work on non-generic files as well, but since this issue is specifically talking about enabling it files that do not have a normal ruleset associated with it, I am sticking to that context.

```js
    "generic": {
        "allow_scanning": false,
        "scan_scopes": ["-comment -string"]
    },
```

Now when we are in a plain text file, for instance, we can force scanning, force scanning to be disabled, or remove our override entirely.

![override](https://user-images.githubusercontent.com/1055125/94275666-753b6400-ff04-11ea-917f-d64e7e89bbbf.gif)

**Comment by facelessuser** [comment]
This work is done and will be included by #158. Usable beta should be available this weekend as there is only one open issue left that is planned.

Closing issue as work for this issue has finally been addressed.
